[clang-tidy] Add option to ignore macros in `readability-simplify-boolean-expr` check

SimplyDanny opened the pull request:
Closes #57975.

SimplyDanny committed:
[clang-tidy] Add option to ignore macros in `readability-simplify-boolean-expr` check

SimplyDanny commented on clang-tools-extra/test/clang-tidy/checkers/readability/simplify-boolean-expr-macros.cpp:
@njames93: The test currently fails. You likely see immediately why. I don't have a clue at the moment. Feels like I tried everything already. 😶

PiotrZSL commented on clang-tools-extra/docs/ReleaseNotes.rst:
```diff
@@ -500,6 +500,10 @@ Changes in existing checks
   <clang-tidy/checks/readability/static-accessed-through-instance>` check to
   identify calls to static member functions with out-of-class inline definitions.
 
+- Added option `IgnoreMacros` to :doc:`readability-simplify-boolean-expr
```
follow current style of change notes. and put checks in alphabetical order

PiotrZSL commented on clang-tools-extra/test/clang-tidy/checkers/readability/simplify-boolean-expr-macros.cpp:
```diff
@@ -0,0 +1,20 @@
+
+// Ignore expressions in macros.
+// RUN: %check_clang_tidy %s readability-simplify-boolean-expr %t
+// RUN:     -- -config="{CheckOptions: [{key: readability-simplify-boolean-expr.IgnoreMacros, value: true}]}"
```
follow new style of options (no need for value or key, check other checks)

PiotrZSL commented on clang-tools-extra/clang-tidy/readability/SimplifyBooleanExprCheck.cpp:
```diff
@@ -277,6 +277,9 @@ class SimplifyBooleanExprCheck::Visitor : public RecursiveASTVisitor<Visitor> {
   }
 
   bool dataTraverseStmtPre(Stmt *S) {
+    if (Check->IgnoreMacros && S->getBeginLoc().isMacroID()) {
```
what if S is nullptr ? it will crash, move S verification from line 283 up.

SimplyDanny committed:
Check for potential `null` value

SimplyDanny committed:
Move entry in release notes

SimplyDanny committed:
Use new option style

SimplyDanny commented:
Thanks @PiotrZSL for the review! I've addressed your remarks. Would you [support me](https://github.com/llvm/llvm-project/pull/78043#discussion_r1451556010) with the test as well?

PiotrZSL commented on clang-tools-extra/test/clang-tidy/checkers/readability/simplify-boolean-expr-macros.cpp:
```diff
@@ -0,0 +1,20 @@
+// RUN: %check_clang_tidy -check-suffixes=,MACROS %s readability-simplify-boolean-expr %t
+
+// Ignore expressions in macros.
+// RUN: %check_clang_tidy %s readability-simplify-boolean-expr %t
+// RUN:     -- -config="{CheckOptions: {readability-simplify-boolean-expr.IgnoreMacros: true}}"
```
missing new line escape, add \ at the end of those 2 lines, and tests will pass

SimplyDanny committed:
Add line continuation characters to make tests pass

SimplyDanny committed:
Adapt style of release notes

SimplyDanny commented on clang-tools-extra/docs/ReleaseNotes.rst:
```diff
@@ -492,6 +492,10 @@ Changes in existing checks
   <clang-tidy/checks/readability/non-const-parameter>` check to ignore
   false-positives in initializer list of record.
 
+- Added option `IgnoreMacros` to :doc:`readability-simplify-boolean-expr
```
You mean by using the "Improves ... check" style? Is it okay now?

PiotrZSL commented on clang-tools-extra/docs/ReleaseNotes.rst:
```diff
@@ -492,9 +492,9 @@ Changes in existing checks
-  It makes the check ignore boolean expressions passed into macros.
+- Improved :doc:`readability-simplify-boolean-expr
+  <clang-tidy/checks/readability/simplify-boolean-expr>` check by adding the
+  new option `IgnoreMacros` that ignores boolean expressions passed into macros.
```
that controls ignoring of expressions that originate from macros.
or something like that.

PiotrZSL reviewed (approved):
LGTM

SimplyDanny committed:
Improve release notes

SimplyDanny committed:
Improve option description
